enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -1793,6 +1804,199 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
 	return pebble.ReadCloserWithFunc(rc, db.Close), nil
 }
 
+// A writer that will chunk bytes written to it using Content-Defined Chunking,
+// and then, if configured, encrypt and compress the chunked bytes.
+type cdcWriter struct {
+	ctx        context.Context
+	pc         *PebbleCache
+	db         pebble.IPebbleDB
+	fileRecord *rfpb.FileRecord
+	key        filestore.PebbleKey
+
+	shouldCompress bool
+	isCompressed   bool
+
+	chunker       *chunker.Chunker
+	writtenChunks []*rspb.ResourceName
+	numChunks     int
+	firstChunk    []byte
+
+	eg *errgroup.Group
+}
+
+func (p *PebbleCache) newCDCCommitedWriteCloser(ctx context.Context, fileRecord *rfpb.FileRecord, key filestore.PebbleKey, shouldCompress bool, isCompressed bool) (interfaces.CommittedWriteCloser, error) {
+	db, err := p.leaser.DB()
+	if err != nil {
+		return nil, err
+	}
+
+	eg, ctx := errgroup.WithContext(ctx)
+	eg.SetLimit(10)
+	cdcw := &cdcWriter{
+		ctx:            ctx,
+		eg:             eg,
+		pc:             p,
+		db:             db,
+		key:            key,
+		fileRecord:     fileRecord,
+		shouldCompress: shouldCompress,
+		isCompressed:   isCompressed,
+	}
+
+	chunker, err := chunker.New(ctx, p.averageChunkSizeBytes, cdcw.writeChunk)
+	if err != nil {
+		return nil, err
+	}
+	cdcw.chunker = chunker
+
+	cwc := ioutil.NewCustomCommitWriteCloser(cdcw)
+	cwc.CloseFn = db.Close
+	cwc.CommitFn = func(bytesWritten int64) error {
+		if err := cdcw.Close(); err != nil {
+			return err
+		}
+		if err := cdcw.eg.Wait(); err != nil {
+			return err
+		}
+		if cdcw.numChunks == 1 {
+			// When there is only one single chunk, we want to store the original file
+			// record with the original key instead of computed digest from the
+			// chunkData. This is because the chunkData can be compressed or encrypted,
+			// so the digest computed from it will be different from the original digest.
+			return cdcw.writeRawChunk(cdcw.fileRecord, cdcw.key, cdcw.firstChunk)
+		}
+		now := p.clock.Now().UnixMicro()
+
+		md := &rfpb.FileMetadata{
+			FileRecord:      fileRecord,
+			StorageMetadata: cdcw.Metadata(),
+			StoredSizeBytes: bytesWritten,
+			LastAccessUsec:  now,
+			LastModifyUsec:  now,
+		}
+		return p.writeMetadata(ctx, db, key, md)
+	}
+	return cwc, nil
+}
+
+func (cdcw *cdcWriter) writeChunk(chunkData []byte) error {
+	cdcw.numChunks++
+
+	if cdcw.numChunks == 1 {
+		// We will wait to write the first chunk until either cdcw.Commit() is
+		// called or the second chunk is encountered.
+		// In the former case, there is only one chunk, we don't want to write a
+		// file-level metadata entry and a chunk entry into pebble.
+		cdcw.firstChunk = make([]byte, len(chunkData))
+		copy(cdcw.firstChunk, chunkData)
+		return nil
+	}
+
+	if cdcw.numChunks == 2 {
+		if err := cdcw.writeChunkWhenMultiple(cdcw.firstChunk); err != nil {
+			return err
+		}
```

Comment:
Should we clear out `cdcw.firstChunk` here? I believe the data is no longer needed after this.

Reply:
Done